Add IO::Socket::SSL and Net::SSLeay to the Perl core

atrodo opened the pull request:
On p5p, the question was asked what would it look like to just add IO::Socket::SSL and Net::SSLeay to the core. It appears that all that was needed is to include the distribution in cpan/, add information to Porting/Maintainers.pl, and include an explicit dependency between the two packages in Makefile.SH.

The tests all pass for me except for cmp_version.t. I think that's because it is confused by the newly added .c/.h files and thinking the versions need to be bumped instead of recognizing them as new. I am hoping to dig more into what is causing that failure and either fixing the files to pass or find a fix for cmpVERSION.pl.

demerphq commented:
JFDI++ 

FWIW think the concern is what to do if they don't or can't build on a platform. We target a lot of OSes and architectures, so the devil will be making sure that what works generally on *nix works ok on our other build targets. 

But this is a great start. Thank you for the effort!

haarg commented:
Good to have a start on this.

A few things that will need to be resolved.

  - There are various extra files that should be excluded, like `cpan/Net-SSLeay/t/local/kwalitee.t`, `cpan/Net-SSLeay/t/local/01_pod.t`, `cpan/Net-SSLeay/t/local/02_pod_coverage.t`, and probably most of the files listed in `Maintainers.pl`'s `@IGNORABLE`.
  - Net::SSLeay is licensed as Artistic 2. I'm not sure if that presents a problem.
  - The probing for openssl headers probably needs to happen in Configure rather than Net::SSLeay's `Makefile.PL`.

jkeenan commented:
> FWIW think the concern is what to do if they don't or can't build on a platform. We target a lot of OSes and architectures, so the devil will be making sure that what works generally on *nix works ok on our other build targets.
> 

I share @demerphq's concerns in this matter.  Unfortunately, the CPAN maintenance of both of these libraries is not currently sufficiently robust to support their redistribution in the Perl core distribution.

When I say "the CPAN maintenance" of these libraries, I'm referring to two things:  (i) the support provided by the people currently maintaining them upstream on CPAN; and (ii) our ability to evaluate how well these libraries are performing across various OSes and architectures.  Let's take each in turn.

**Upstream Support**

When we decide to include a CPAN distribution in core Perl, we are adding both a technical dependency and a human dependency.  The technical dependency (*e.g.,* which files in the distribution need to be omitted from core) has already begun to be addressed in this pull request by @atrodo, @Leont and @haarg.  The human dependency remains to be addressed.  Do the current upstream maintainers want their distributions to be shipped with core?  Are the upstream maintainers willing to take on the extra work associated with inclusion of their libraries in the Perl core?

Let's first acknowledge two things about Net-SSLeay and IO-Socket-SSL.  First, the subject matter of these libraries is very complex -- probably more complex than any other libraries we ship with core.  If you're not familiar with these complexities, then you should pause reading this post right now and read at least these parts of their documentations:

[Net-SSLeay Compatibility](https://metacpan.org/pod/Net::SSLeay#COMPATIBILITY)

[IO-Socket-SSL Description](https://metacpan.org/pod/IO::Socket::SSL#DESCRIPTION)

Second, the current and past maintainers of these distributions do put a lot of work into maintaining them.  Both Net-SSLeay and IO-Socket-SSL have had new releases in the past month.  There are libraries we ship with core Perl which have not had new releases in years.

Nonetheless, the sheer complexity of these libraries may mean that their upstream maintainers have to make decisions about the scope of their support that would unfavorably impact the value of having these libraries in core.  Let me give one example from my own experience.

*Blead Breaks CPAN*

A large part of the QA work that I and other P5Pers do is to assess the impact of changes in development versions of Perl on CPAN libraries -- the so-called "Blead Breaks CPAN" (BBC) problem.  We rely upon twenty years of work on the Perl toolchain to enable people to file CPANtesters reports.  We rely upon other people to analyze those reports and file bug reports when blead breaks CPAN.  Since these bugs are often platform-specific, we rely upon our ability to install toolchain modules such as Task::CPAN::Reporter on all platforms we actively support.  More specifically, we need to be able to install these toolchain modules on all platforms on development versions of Perl (5.37.5, 5.37.6, etc.) -- not just on the last production release of Perl (5.36.0).  If we can't install the toolchain modules on development versions of Perl on a specific supported platform, then we can't fully assess the impact of changes in blead on CPAN libraries subsequently installed on that platform.

*IO-Socket-SSL not installable against dev versions of perl on OpenBSD*

Last February I wanted to install Task::CPAN::Reporter on OpenBSD so that I could generate CPANtesters reports on that platform against blead in the leadup to the release of perl-5.36.0.  I was unable to do so because Task::CPAN::Reporter has a dependency on IO::Socket::SSL and IO::Socket::SSL's tests were failing, rendering it (and its reverse dependencies) not installable *from CPAN.*  I filed a [bug ticket](https://github.com/noxxi/p5-io-socket-ssl/issues/115) on GitHub about this problem.  The maintainer of IO-Socket-SSL responded, "Unfortunately LibreSSL [used by OpenBSD] is a kind of moving target with varying support for TLS 1.3 and different API support in different versions of LibreSSL. This makes it very hard to support IO::Socket::SSL and Net::SSLeay on LibreSSL."  We both acknowledged that Task::CPAN::Reporter could be installed against the *system* `perl` on OpenBSD because "[t]he OpenBSD ports for Net::SSLeay and IO::Socket::SSL ... come with their own patches - which sometimes skip tests and sometimes enforce TLS 1.2."  But to adequately respond to BBC reports we need to be able to install toolchain modules against monthly dev versions of `perl` and not just upon the system `perl` and packages updated (in the case of OpenBSD) every six months.

The maintainer closed my ticket.  I have been unable to file CPANtesters reports against dev versions of `perl` on OpenBSD since then -- and I confirmed this problem by trying to install IO::Socket::SSL today.

**Library Performance**

Apart from my own particular problems with IO-Socket-SSL and Task::CPAN::Reporter on one particular OS, we are currently not getting data sufficient to judge the performance of either Net-SSLeay or IO-Socket-SSL to assess how well they are doing against development versions of `perl`.

*Net-SSLeay* [Net-SSLeay on CPANtesters](http://fast-matrix.cpantesters.org/?dist=Net-SSLeay%201.92)

The latest production release of this library against dev versions of `perl` is largely failing on NetBSD.  Reports on OpenBSD are being graded `UNKNOWN`.  We have had no CPANtesters reports on MSWin32 since perl-5.32.1.

*IO-Socket-SSL* [IO-Socket-SSL on CPANtesters](http://fast-matrix.cpantesters.org/?dist=IO-Socket-SSL%202.079)

The latest production release of this library (2.080) has only been out five days, so let's look at the previous release (2.079) instead.  That version has not been tested on either MSWin32 or OpenBSD since perl-5.32.1.  It has never been tested on NetBSD.  We don't have any reports on dev releases of `perl` from the current (5.37) cycle on cygwin or solaris.

Before we release a dev version of `perl` each month, we often run a program shipped with core, `Porting/core-cpan-diff`, to determine which `cpan/` modules have been updated and need "synching" into core.  Before we do that synching, we try to look at CPANtesters reports of those modules against blead to spot failures which would cause us to put off synching those modules into blead until such failures can be corrected.  But that assumes we have CPANtesters data to work with.  In the case of the two distributions proposed for inclusion in core, we don't currently have sufficient data.

**Summary**

The maintenance of these two libraries is very challenging.  The upstream maintainers have largely met that challenge, but it is their prerogative to set limits on the scope of their maintenance work.  Unfortunately, inclusion in the core would mean an even larger scope in what we need those libraries to do, and we would need much more testing of them to ensure their performance.  I have no objection to the goal of "SSL support out of the box," but these two distributions are not yet in good enough shape to be shipped with core.

atrodo commented:
I hadn't thought about it being in two commits and I think that would be better, so I'll split those into their own commits and add the exclusions. I'll also look at making it optional and at moving the probing into Configure, but I'm going to guess that changing Configure may end up being over my head.

As to the questions of being able to include these two modules because of licenses or support, I will sit those conversations out as I'm not in position to advocate or challenge those questions.

atrodo committed:
Add Net::SSLeay to the Perl core

atrodo committed:
Handle new modules as if the version was bumped

Checking that a module version was bumped when .c/.h files are updated had
an edge case where the previous version was unknown, notably when adding a
new module to cpan/.

atrodo committed:
Add IO::Socket::SSL to the Perl core

atrodo committed:
Do not attempt Net::SSLeay if the header is missing

atrodo committed:
Skip Net::SSLeay if building with C++

Currently, it appears that Net::SSLeay will not compile if perl was compiled
with C++, so skip it on C++ builds for now

leonerd commented:
I agree that this wants to be split into two smaller commits but to keep things moving I've approved the CI run to at least test what we have so far. If that doesn't reveal any big issues, we can look into splitting into two PRs.

haarg commented:
Does anyone know the answer to the licensing issue? Net::SSLeay does not use the standard perl license.

demerphq commented:
> Does anyone know the answer to the licensing issue? Net::SSLeay does not use the standard perl license.

@codesections and @StuartJMackintosh addressed prior legal issues in our code in https://github.com/Perl/perl5/pull/19163 maybe they can do the same here.

demerphq commented:
> to keep things moving I've approved the CI

If @atrodo separates out the AUTHORS/.mailmap patch to an individual patch, and pushes it as a separate PR, maybe including a typo fix or comment tweak or something like that to justify the merge (not strictly necessary but probably "Good form"), then we can approve and merge that and then CI wont need approval (afaik) for any further work he does on this ticket. He only requires approval for CI because this is his first PR and he isn't "trusted" yet. (Note the important part is having a commit merged, not that its to AUTHORS/.mailmap, but that is an easy route to getting him "into the club" so to speak.)

qbit commented:
@h-vn, @mikemccauley, @chrisnovakovic  sorry for the ping but it seems like the most direct way to reach out :D 

If these libraries were moved to core, how would it impact things from your standpoint? Is it something you want?

Re the licensing side of things.. there is currently [this](https://github.com/Perl/perl5/blob/blead/cpan/Encode/bin/encguess#L141) file in core which is Artistic License (2.0) (and possibly others, I haven't looked too hard).

atrodo commented:
I am going to attempt to summarize where I think this PR is at:
* This should be split into 2 PRs, one for Net::SSLeay and one for IO::Socket::SSL
  * When Leont said two commits, I assumed he meant two actual commits since it was originally 1 commit. It now sounds like two PRs were meant
  * If this PR gets the greenlight on the other items, I will split the changes for IO::Socket::SSL into a new PR
* I will look to create a separate PR for the patch to Porting/cmpVERSION.pl
  * Thinking more about it, it probably deserves its own PR
  * This would be a good change for demerphq's suggestion to get the CI to run without approval
* There are issues with Windows
  * These modules should be skipped if the headers are not found, but I'm unfamiliar with the windows build process, so that will take some research
* Net::SSLeay has more strict version requirements than are reflected here
  * The current Configure changes merely look for the openssl header, and does no version checking
  * How that is done different enough code cannot be shared, Configure is in sh, and the module is rightfully running perl code
  * I think this will have to be handled by doing similar version compares in Configure, but I have concerns about the ability to maintaining the compatibility list between the two systems
* I have thought about adding an option to skip installing SSL support
* There is an open issue on Net::SSLeay to make compiling with C++ possible
  * If this is closed with a new release, I will update and remove the C++ skip code
* There is some discussion around if the maintainers are receptive to these modules being included
* There is also some question around if perl wants to include these two modules
* There are some questions about license compatibility

Those last 3 bullets must be addressed before this PR can be considered for merging, and those 3 things are largely out of my capacity to answer or push forward. In the meantime, I will work on the technical items as I can.

mikemccauley commented:
Hello Aaron and All,

On Saturday, 18 February 2023 05:02:42 AEST Aaron Bieber wrote:
> @h-vn, @mikemccauley, @chrisnovakovic  sorry for the ping but it seems like
> the most direct way to reach out :D
> 
> If these libraries were moved to core, how would it impact things from your
> standpoint? Is it something you want?

I have no objection.

Cheers.

> 
> Re the licensing side of things.. there is currently
> [this](https://github.com/Perl/perl5/blob/blead/cpan/Encode/bin/encguess#L1
> 41) file in core which is Artistic License (2.0) (and possibly others, I
> haven't looked too hard).


-- 
Mike McCauley           VK4AMM                   ***@***.***
Airspayce Pty Ltd 9 Bulbul Place Currumbin Waters QLD 4223 Australia   
http://www.airspayce.com                                 5R3MRFM2+X6
Phone +61 7 5598-7474

Leont commented:
> * This should be split into 2 PRs, one for Net::SSLeay and one for IO::Socket::SSL 
>   * When Leont said two commits, I assumed he meant two actual commits since it was originally 1 commit. It now sounds like two PRs were meant
>   * If this PR gets the greenlight on the other items, I will split the changes for IO::Socket::SSL into a new PR

No I meant two commits. I'm fine with them coming in one PR as `IO::Socket::SSL` should be trivial to add once you got `Net::SSLeay`.

> * I will look to create a separate PR for the patch to Porting/cmpVERSION.pl
>   * Thinking more about it, it probably deserves its own PR
>   * This would be a good change for demerphq's suggestion to get the CI to run without approval

Yeah it would make sense to split that out.

> There are issues with Windows
> *  These modules should be skipped if the headers are not found, but I'm unfamiliar with the windows build process, so that will take some research

Windows is typically configured by editing the constants in the relevant makefile, I believe.

> * Net::SSLeay has more strict version requirements than are reflected here
>   * The current Configure changes merely look for the openssl header, and does no version checking
>   * How that is done different enough code cannot be shared, Configure is in sh, and the module is rightfully running perl code
>   * I think this will have to be handled by doing similar version compares in Configure, but I have concerns about the ability to maintaining the compatibility list between the two systems

Yeah, that would be where most of the complexity of this PR would be.

chrisnovakovic commented:
> @h-vn, @mikemccauley, @chrisnovakovic sorry for the ping but it seems like the most direct way to reach out :D
> 
> If these libraries were moved to core, how would it impact things from your standpoint? Is it something you want?

Apologies for the slow reply - @h-vn and I have been speaking offline about this for the last week.

We appreciate the reasoning behind this effort, and we also appreciate how this will make life easier for the majority of both Perl developers and end users. Notwithstanding the time Heikki and I (as well as numerous other contributors) have put into modernising and improving platform support for Net-SSLeay since we took over as maintainers five years ago, it remains true that the installation process is sometimes a frustrating experience, and including it in core would make that problem go away for the most part.

Having said that, we have some concerns. In no particular order:

* We're unsure about the additional maintenance burden that this would place on us. I maintain Net-SSLeay in my (limited) spare time; Heikki maintains it as part of his employment with Radiator Software, although it isn't his full-time focus. We're reasonably responsive to PRs, bug reports and support requests, and we're usually diligent about adding support for forthcoming releases of OpenSSL and LibreSSL, but I don't know how much that effort would become expected of us, or whether we could guarantee that level (or perhaps even a greater level) of support.
* Making Net-SSLeay dual-life presumably means that it suddenly has the capability to block new stable releases of Perl if it doesn't work as expected with a new stable release of OpenSSL or LibreSSL that makes its way into an OS that Perl targets. API compatibility has caused problems for us in the recent past, and there aren't always quick and/or easy solutions. For example, Net::SSLeay didn't work with the first two stable releases of LibreSSL 3.2 (which was included in OpenBSD 6.8) because they contained a new X.509 verifier that wasn't bug-compatible with libssl; we had to wait four months for LibreSSL 3.2.4, which reverted to LibreSSL 3.1's libssl-compatible verifier, before Net::SSLeay worked on OpenBSD 6.8. There's also the case that @jkeenan alluded to, in which LibreSSL's TLS 1.3 implementation (introduced in version 3.1.1) didn't provide a fully libssl-compatible API until version 3.4.1, so Net-SSLeay appeared to be broken on OpenBSD for a period of about 18 months. This isn't intended as a criticism of the LibreSSL maintainers - TLS is a complex protocol, libssl is a complex API, and they've embarked upon a complex endeavour by trying to implement the former while remaining bug-compatible with OpenSSL's implementation of the latter. It's just that there's nothing that could realistically have been done in Net-SSLeay to make it work in either of these cases, and we're concerned about the possibility that it could have held up a stable Perl release had it been part of core at the time.
* The high level of configurability of libssl (via compile-time options and run-time settings) means that there's a bewildering number of possible setups even within the same libssl version, and Net-SSLeay can't be compatible with all of them. For example, some vendors that package OpenSSL deviate from the default compile-time behaviour by disabling deprecated functions - Net-SSLeay has to use some deprecated functions to implement certain features that are expected to work across the full range of supported OpenSSL versions, so our position is simply that Net-SSLeay isn't compatible with the OpenSSL packages provided by those vendors. "Build a compatible libssl yourself" may be a tenable position for a module on CPAN, but perhaps it isn't for a core module.
* With a very small number of exceptions, Net-SSLeay supports every stable release of OpenSSL going back to 0.9.8c and every stable release of LibreSSL there has ever been (with the usual caveats around depending on any out-of-support libssl implementation without also getting security patches from a third-party vendor). We also intentionally aligned Net-SSLeay's Perl version support policy with that of the [Lancaster Consensus](https://github.com/Perl-Toolchain-Gang/toolchain-site/blob/master/lancaster-consensus.md), both because it seemed like the responsible thing to do for a distribution containing a river stage 4 module, and in anticipation of Net-SSLeay eventually becoming dual-life. However, the libssl support policy is becoming increasingly untenable - the changes made in OpenSSL 3.0 in particular are making it difficult to support both OpenSSL 0.9.8 and 3.0 simultaneously, and we may therefore be forced to drop support for older OpenSSL versions at some point. How would such a change square with a requirement for core modules to support old Perl versions, given that use of an old Perl usually implies use of an equally old libssl? Here are two concrete examples: if Windows users need to use Perl 5.8, they'll probably download Strawberry Perl 5.8.9.5, which bundles OpenSSL 0.9.8l. Is Net-SSLeay expected to be installable from CPAN in that environment? RHEL 5 users are using Perl 5.8.8 and OpenSSL 0.9.8e out of the box. Is Net-SSLeay expected to be installable from CPAN in that environment?
* Net-SSLeay dates back to 1996, and there are features in it that we're keen for people not to use for much longer. Good examples of this are Net::SSLeay::Handle and the [high-level API in Net::SSLeay](https://metacpan.org/pod/Net::SSLeay#High-level-functions-for-accessing-web-servers), which exposes a number of methods for opening TLS sockets and talking to HTTPS (and even HTTP) servers: these are obviously jobs for IO::Socket::SSL, HTTP::Tiny, LWP, etc in 2023. We haven't started addressing this yet, partially through lack of time to devote to cleaning up tech debt and partially due to the care that needs to be taken when deprecating and removing features that have been in a river stage 4 module for 25+ years. We don't want to give the impression that p5p is in some way condoning the use of these features just because they're there in a core module - we remember CGI.pm being given the boot from core for exactly that reason.
* The Net::SSLeay documentation is pretty out-of-date, and to be honest we don't think it currently meets the standard of quality we'd expect of documentation for a core module. Again, this is something we plan to fix, but there's a lot of work to do and not a lot of spare time in which to do it.

It would also be good to get a perspective from @noxxi, as the maintainer of IO-Socket-SSL.

noxxi commented:
> It would also be good to get a perspective from @noxxi, as the maintainer of IO-Socket-SSL.

My concerns about this are similar.

If the distributions move into core I would have higher standards on the production quality than if they stay out of core. Production quality here means less the coding and documentation but more the timely support, i.e. having sufficient knowledgeable developers to test the software, react to user questions, align shipping deadlines with core release times ... But   I'm currently the single maintainer of IO::Socket::SSL and cannot allocate too much resources (both work and spare time) to it - this is not a question of pay but more of many other things to do. And debugging any kind of SSL problems is very hard.

There are also problems at the coding level: while I have some review from coworkers and we use IO::Socket::SSL for production, there is not much testing and review done outside of it. This means the focus of testing is on a very few platforms and a limited set of functionality. Platform support for IO::Socket::SSL means not only the OS itself, but the already mentioned broad variety of OpenSSL versions, forks like LibreSSL with their own "personality" and also different versions of Net::SSLeay.  While both Net::SSLeay and IO::Socket::SSL come with many tests, they are far from what is needed to actually cover lots of edge cases which happen in practice.

On the positive side: Moving IO::Socket::SSL and Net::SSLeay to the  core might add more stable dependencies or at least have a clearer focus on which dependencies (OpenSSL versions) need to be well supported and which combinations happen only if the user insists of using their own dependencies.

In summary: before moving IO::Socket::SSL into core it should be made more robust in terms of developers involved in the project and tests done with the diversity of dependencies.

mbeijen commented:
> Here are two concrete examples: if Windows users need to use Perl 5.8, they'll probably download Strawberry Perl 5.8.9.5, which bundles OpenSSL 0.9.8l. Is Net-SSLeay expected to be installable from CPAN in that environment? RHEL 5 users are using Perl 5.8.8 and OpenSSL 0.9.8e out of the box. Is Net-SSLeay expected to be installable from CPAN in that environment?

While I absolutely admire all the work you do as a software maintainer, I really think you make it very difficult for yourself here! First of all, this discussion here is about bringing Net-SSLeay into core perl. So it does not concern perl 5.8, it concerns perl 5.38, there is more than 20 years between those versions.

Second, if I would somehow want to use Perl 5.8 I definitely do not expect I can use it with the newest version of OpenSSL or conversely that I can expect to install the newest version of Net-SSLeay with my very old OpenSSL.

For your example about 'would RHEL5 users expect to be able to install Net-SSLeay on perl 5.8.8?'
 - If you're using the system perl, you can install the net-ssleay module from the repos or archives or the CD-rom it came on and that would work. Then you'll have an ancient perl with matching ancient Net::SSLeay and libssl
 - If you want a new perl, surely you can bring your own SSL libs as well?
 - For fun and giggles I got a CentOS 6 (so based on RHEL *6*) machine and tried to compile modern Python, as that also comes with SSL bindings. But I was not able to install a C compiler, as the repolist no longer works, because it's no longer supported, because it's too old... and RHEL 5 pre-dates that release by another five years! 

and to answer the question, if I would be a RHEL 5 user I would be amazed if I would be able to install any amount of new software. I also can't install new PostgreSQL versions, new nginx, most other software or at least not without an extreme amount of trouble.
 
I really would say, especially from the perspective of perl 5 development, we would not need to care that there will be no Net::SSLeay if you somehow really want to compile modern perl 5 on such an old machine.

Please also note that ssl is a security module, you might not even _want_ to let users compile very old and insecure versions of openSSL with modern Net::SSLeay

See also https://peps.python.org/pep-0644/ where they explicitly decide in 2021 to drop support for all openSSL versions before 1.1.1 because it's just _too much_ maintenance work for them!

I think this underscores how amazing all your maintenance work is, that you seem to be able to pull this off (with caveats) while the Python community can't.

Of course, perl is not python!

bulk88 commented:
(very long thread and goes offtopic sometimes) https://www.nntp.perl.org/group/perl.perl5.porters/2021/12/msg262180.html

https://github.com/perl-libwin32/win32/pull/30

Why was all of this work done for Perl on WinOS (basic CURL from MS's Core OS native no-GUI ever/server grade HTTP library in a  DLL), yet zero hits on metacpan, for its usage 2 years later?

bulk88 commented:
I just noticed ```Net::SSLeay```  ver 1.92  **DOES NOT** ```#define PERL_NO_GET_CONTEXT```.  Terrible for performance. That module should not be in perl core until at min that is fixed on CPAN.

It is 2024. If XS CPAN modules don't have PERL_NO_GET_CONTEXT after 2 decades of perl threads, that shows alot of technical debt or very low free time by that module's author.

Leont commented:
> I just noticed Net::SSLeay ver 1.92 DOES NOT #define PERL_NO_GET_CONTEXT. Terrible for performance. That module should not be in perl core until at min that is fixed on CPAN.

Yeah, it would need a thorough cleanup.

bulk88 commented:
https://metacpan.org/release/SULLR/IO-Socket-SSL-2.095/source/lib/IO/Socket/SSL/Intercept.pm#L41

I can't track down the RV2CV optimization thread from last year, I wanna attack some of the BBCs ATM, but 2-3 of the CPAN modules that got broken by RV2CV last year, have problems like these output being weird/wrong/broken/different than Perl from 5 years ago
```
C:\sources\perl5\win32>perl -E" say ref(\*Internals::V)"
GLOB
C:\sources\perl5\win32>perl -E" say ref(\*Internals::V{CODE})"
REF
C:\sources\perl5\win32>perl -E" say ref(*Internals::V{CODE})"
CODE
C:\sources\perl5\win32>
```
Now imagine, in a fictional scenario, if RV2CV optimization alters the behavior of this line [IO/Socket/SSL/Intercept.pm](https://metacpan.org/release/SULLR/IO-Socket-SSL-2.095/source/lib/IO/Socket/SSL/Intercept.pm#L41)

https://metacpan.org/release/SULLR/IO-Socket-SSL-2.095/source/lib/IO/Socket/SSL/Intercept.pm#L134 Eghhhh, Storable has alot of bit rot, Ive tried 3x to refactor it in the last 18 months. Storable's CPU burn is the #1 reason why `cpan.pl` shell takes 3-6 seconds to start up.
